### ISSUE TITLE
SQUASHFS error: failed to initialize zstd decompressor

### ISSUE THREAD

**Issue by tpgxyz** [facebook/zstd#799]
Hi,

i've managed to import your zstd patches v5 into kernel 4.12.x and into squashfs-tools.

Unfortunately ISO build with these does not boot.
https://discourse-static.openmandriva.org/uploads/default/optimized/2X/8/8e25828c315fa819117dd0c4e4f6d2ea7d447e8a_1_690x388.png

ISO is compressed with squashfs-tools with zstd compression as default.

You can download and try this ISO  from here https://abf.openmandriva.org/platforms/3.0/products/38/product_build_lists/1358

suqshfs-tools: https://github.com/OpenMandrivaAssociation/squashfs-tools
kernel: https://github.com/OpenMandrivaAssociation/kernel-release

If you need any more information please give a sign.

**Comment by terrelln** [comment]
I can't quite read the errors, does it say something along the lines of:

```
[ 4377.243871] SQUASHFS error: Failed to initialize zstd decompressor
[ 4377.243872] SQUASHFS error: zstd decompression failed, data probably corrupt
[ 4377.243873] SQUASHFS error: squashfs_read_data failed to read block 0x28b3d
[ 4377.243874] SQUASHFS error: Unable to read metadata cache entry [28b3d]
[ 4377.243875] SQUASHFS error: Unable to read inode 0x106709e0
```

v4 of the patches had this issue, can you confirm that `fs/squashfs/zstd_wrapper.c:82` is this?


```c
stream = ZSTD_initDStream(wksp->window_size, wksp->mem, wksp->mem_size);
```

**Comment by tpgxyz** [comment]
@terrelln 
Yes kernel v5 patch does have this line
https://github.com/facebook/zstd/blob/dev/contrib/linux-kernel/0004-squashfs-Add-zstd-support.patch#L235

And my kernel is compiled with v5 zstd patches from your official repository.

**Comment by tpgxyz** [comment]
Here is the iso with latest kernel patches v5 and squashes-tools patch v6.
Unfortunately it does not boot with same error.

https://abf.openmandriva.org/platforms/3.0/products/38/product_build_lists/1361

Help?

**Comment by terrelln** [comment]
@tpgxyz I will try to reproduce the error. It does sound like #784, but I fixed that particular bug in v5.

**Comment by terrelln** [comment]
Switching squashfs-tools to v6 was a comments only change and doesn't effect the ability of the squashfs in the kernel to initialize the decompression context.

Does it consistently work with the latest version of the patches, or are there consistent or sporadic failures?

**Comment by tpgxyz** [comment]
@terrelln As I wrote Iso does not boot with v5 kernel patches and with squashfs-tools v6 patch.

**Comment by terrelln** [comment]
Oh sorry I parsed that incorrectly, I read it as it does not have the same error when booting. Thanks for the clarification.

**Comment by terrelln** [comment]
@tpgxyz I think I've found the issue. I haven't been able to reproduce the failure, so would you be able to check that it works? The short workaround is on `zstd_wrapper.c:46`:

```c
	wksp->mem_size = ZSTD_DStreamWorkspaceBound(wksp->window_size) + sizeof(size_t);
```

I think the problem occurs when `vmalloc()` returns a value that isn't aligned to a `size_t`. If that fixes that for you, I will fix the `ZSTD_XWorkspaceBound()` functions.

**Comment by terrelln** [comment]
You'll also have to change `lib/zstd/zstd_common.c:40` to:

```c
	if (!workspace || workspaceSize < sizeof(ZSTD_stack)) {
```

**Comment by terrelln** [comment]
@tpgxyz I've tested on on v4.12 with gcc-4.8.5 and gcc-7 (with your btrfs-zstd backport), and v4.13-rc5 with the same compilers as well as clang-5.0 (with patches applied from https://github.com/ramosian-glider/clang-kernel-build), and haven't been able to reproduce the issue.

My previous guess at the issue shouldn't be possible because `vmalloc()` should return page aligned memory, but it is the only thing that I can think of (unless it is UB).

The data being decompressed shouldn't matter, since the failure is when initializing the `ZSTD_DStream`. The only two things that can cause failures in this part of the code are not enough memory in the workspace, or the workspace is misaligned. I've triple checked that `ZSTD_DStreamWorkspaceBound()` is large enough.

I'm building the kernel with a very different configuration, with zstd/squashfs as modules, but I don't expect it to fail. After that, there isn't too much more I can do on my side, since I can't reproduce it. I would be interested in verifying that `wksp` is as expected, could you run with this patch

```patch
 fs/squashfs/zstd_wrapper.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/fs/squashfs/zstd_wrapper.c b/fs/squashfs/zstd_wrapper.c
index eeaabf8..6b28a28 100644
--- a/fs/squashfs/zstd_wrapper.c
+++ b/fs/squashfs/zstd_wrapper.c
@@ -79,6 +79,9 @@ static int zstd_uncompress(struct squashfs_sb_info *msblk, void *strm,
        ZSTD_inBuffer in_buf = { NULL, 0, 0 };
        ZSTD_outBuffer out_buf = { NULL, 0, 0 };

+       printk("wksp->mem = %p\n", wksp->mem);
+       printk("wksp->mem_size = %zu\n", wksp->mem_size);
+       printk("wksp->window_size = %zu\n", wksp->window_size);
        stream = ZSTD_initDStream(wksp->window_size, wksp->mem, wksp->mem_size);

        if (!stream) {
```

and let me know what it prints? I get

```
wksp->mem = ffffb57640115000
wksp->mem_size = 549424
wksp->window_size = 131072
```

**Comment by tpgxyz** [comment]
@terrelln Hi, I'm on holidays, so I'm unable to test your patch.

Calling @bero for a help.

**Comment by itchka** [comment]
@terrelln I've picked this up while @tpgxyz  and @berolinux are away I have applied the patch above. Most strangely everything now works.
Here are the numbers 
wksp->mem = ffffb57640115000
wksp->mem_size = 549424
wksp->window_size = 131072
Which are the same as yours.
So adding that code seems to initialise the decompressor. It would seem that the error message was correct.
Perhaps we are over optimising the kernel with our settings.
@itchka

**Comment by benbullard79** [comment]
@itchka so can we now build an ISO with zstd/squashfs that boots?

**Comment by itchka** [comment]
Ben the answer to your question is yes though it's not yet a proper fix. I have built an ISO locally and it works so we are heading in the right direction. I am not available until later tonight when I will commit my changes and build a test kernel. Once that's done an ISO can be built and we can at least test the rest of it.

**Comment by terrelln** [comment]
@itchka Does it fail without the extra logging patch?

**Comment by itchka** [comment]
@terrelln I am most confused I have backed out the debug patch now and rebuilt the kernel and embarassingly it works. While I was applying the patch I did add the btrfs-zstd backport and also updated the crypto-add-zstd-support patch where the Makefile patch in it was not applying. These two combined appear to have corrected the problem. I think the backport patch may have been the answer though; I should perhaps have been more methodical. Thank you for all your help on this.
